### PR TITLE
Fix coalescing issue

### PR DESCRIPTION
**Pull request by robertzk**

### CODE REVIEW

File: tests/testthat/test-stage_runner.r [hunk 1]
```diff
@@ -448,3 +448,11 @@ test_that('it correctly returns the furthest executed stage in a stageRunner', {
   expect_identical(sr$current_stage(), '3')
 })
 
+describe("$stage_name1", {
```
Comment (peterhurford) [inline tests/testthat/test-stage_runner.r]: $?

Comment (robertzk) [inline tests/testthat/test-stage_runner.r]: Method name on the stagerunner

**Comment by peterhurford** [comment]
Looks legit.

**Comment by peterhurford** [comment]
...Sorry for accidentally closing. :/

File: tests/testthat/test-stage_runner.r (hunk 1 above)
Comment (kirillseva) [inline tests/testthat/test-stage_runner.r]: dolla dolla bill y'all

File: R/stagerunner-coalesce.R [hunk 2]
```diff
@@ -39,7 +39,7 @@ stageRunner_coalesce <- function(other_runner) {
   # stopifnot(remember)
   if (!isTRUE(self$remember)) return()
 
-  ## We must handle this cases: (1) integration with 
+  ## We must handle these cases: (1) integration with
```
Comment (kirillseva) [inline R/stagerunner-coalesce.R]: attention to details 👍

File: R/stagerunner-stage_names.R [hunk 3]
```diff
@@ -16,7 +16,7 @@
 stageRunner_stage_names <- function() {
   nested_stages <- function(x) {
     if (is.stagerunner(x)) {
-      nested_stages(x$stages)
+      lapply(x$stages, nested_stages)
```
Comment (kirillseva) [inline R/stagerunner-stage_names.R]: nested names?

File: tests/testthat/helper-list_setequal.R [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+#'    and FALSE otherwise.
+list_setequal <- function(list1, list2) {
+  if (is.null(names(list1)) || is.null(names(list2))) {
+    stop("Only named lists can be compared using list_setequal.")
+  }
+  stopifnot(all(nzchar(list1)), all(nzchar(list2)))
+  stopifnot(setequal(names(list1), names(list2)))
+  isTRUE(all.equal(list1, list2[names(list1)] ))
+}
+
```
Comment (kirillseva) [inline tests/testthat/helper-list_setequal.R]: whitespace?

File: tests/testthat/test-printing.r [hunk 5]
```diff
@@ -114,5 +114,53 @@ local({
...
+                          list(foo = list(bar = list(baz = function(e) e$x <- 2, qux = function(e) e$y <- 2)),
+                               second = list(sub1 = function(e) e$z <- 2, sub2 = function(e) e$w <- 2)),
+                          remember = TRUE)
+        }
+
+        r1 <- runner1(); r1$run(to = "1/1/1")
+        r2 <- runner2(); r2$coalesce(r1)
+        expect_output(print(r2), wrapc(" \\* foo [\n\r]*   \\* bar [\n\r]*     \\+ baz [\n\r]*     \\* qux [\n\r]* \\- second [\n\r]*   \\- sub1 [\n\r]*   \\- sub2 ", 4))
+        r2$run("1/1/2")
+        expect_output(print(r2), wrapc(" \\+ foo [\n\r]*   \\+ bar [\n\r]*     \\+ baz [\n\r]*     \\+ qux [\n\r]* \\* second [\n\r]*   \\* sub1 [\n\r]*   \\- sub2 ", 4))
```
Comment (kirillseva) [inline tests/testthat/test-printing.r]: didn't know about wrapc 👍

**Comment by kirillseva** [comment]
LGTM stylewise and I appreciate thorough tests 👍

File: tests/testthat/test-printing.r (hunk 5 above)
Comment (peterhurford) [inline tests/testthat/test-printing.r]: He just custom defined it `wrapc <- function(...) wrap(..., prefix = " caching")`

**Comment by peterhurford** [comment]
Yep, thorough testing is super important here...

File: tests/testthat/test-printing.r (hunk 5 above)
Comment (robertzk) [inline tests/testthat/test-printing.r]: It's a helper I defined earlier, haha.

File: tests/testthat/test-coalescing.R [hunk 6]
```diff
@@ -192,3 +193,67 @@ describe('with tracked_environments', {
   })
 })
 
+test_that("it can coalesce a deeper stagerunner", {
+  test_that("it can coalesce deep runners with vanilla environments", {
+    runner1 <- function() {
+      stageRunner$new(new.env(), list(foo = list(bar = list(baz = function(e) e$x <- 1, qux = function(e) e$y <- 1)),
+                                      second = list(sub1 = function(e) e$z <- 1, sub2 = function(e) e$w <- 1)),
+                      remember = TRUE)
```
Comment (iveksl2) [inline tests/testthat/test-coalescing.R]: yuk

File: tests/testthat/test-coalescing.R [hunk 7]
```diff
@@ -192,3 +193,67 @@ describe('with tracked_environments', {
...
+                      list(foo = list(bar = list(baz = function(e) e$x <- 1, qux = function(e) e$y <- 1)),
+                           second = list(sub1 = function(e) e$z <- 1, sub2 = function(e) e$w <- 1)),
+                      remember = TRUE)
+    }
+    runner2 <- function() {
+      stageRunner$new(objectdiff::tracked_environment(),
+                      list(foo = list(bar = list(baz = function(e) e$x <- 2, qux = function(e) e$y <- 2)),
+                           second = list(sub1 = function(e) e$z <- 2, sub2 = function(e) e$w <- 2)),
+                      remember = TRUE)
+    }
```
Comment (iveksl2) [inline tests/testthat/test-coalescing.R]: closing my eyes

File: .travis.yml [hunk 8]
```diff
@@ -7,7 +7,7 @@ r_github_packages:
   - hadley/devtools
   - smbache/magrittr
   - hadley/testthat
-  - kirillseva/objectdiff@stagerunner_debug_message
+  - robertzk/objectdiff
```
Comment (kirillseva) [inline .travis.yml]: haha :)

**Pull request merged**